face_api_dataset/dataset.py

```diff
@@ -366,37 +372,72 @@ def transform(modalities: List[Modality]) -> List[Modality]:
                 continue
             for modality in modalities:
                 for extension in _modality_files(modality):
-                    if not (self.root / f"{number}.{extension.value}").exists():
+                    if not (self._root / f"{number}.{extension.value}").exists():
                         raise ValueError(
                             f"Can't find file '{number}.{extension.value}' "
                             f"required for {modality.name} modality"
                         )
             image_numbers.add(number)
-        self.image_numbers = sorted(list(image_numbers), key=int)
-        self._image_sizes = {}
-        self.transform = transform
+        self._image_numbers = sorted(list(image_numbers), key=int)
+        self._image_sizes: Dict[str, int] = {}
+        self._transform = transform
 
-    def __len__(self) -> int:
-        return len(self.image_numbers)
+    @property
+    def segments(self) -> Dict[str, int]:
+        """
+        Segment mapping for the dataset.
 
-    def __getitem__(self, i: int) -> dict:
-        if self.transform is None:
-            return self._get(i)
+        :type: Dict[str, int]
+        """
+        return self._segments
+
+    @property
+    def modalities(self) -> List[Modality]:
+        """
+        List of the loaded modalities.
+
+        :type: List[Modality]
+        """
+        return self._modalities
+
+    def __len__(self) -> int:
+        return len(self._image_numbers)
+
+    @overload
+    def __getitem__(self, index: int) -> dict:
+        pass
+
+    @overload
+    def __getitem__(self, index: slice) -> "FaceApiDataset":
+        pass
+
+    def __getitem__(self, i: Union[int, slice]) -> Union[dict, "FaceApiDataset"]:
+        if isinstance(i, slice):
+            ret = copy.copy(self)
+            ret._image_numbers = self._image_numbers[i]
+            ret._image_sizes = {}
+            for key, value in self._image_sizes.items():
+                if key in ret._image_numbers:
+                    ret._image_sizes[key] = value
+            return ret
         else:
-            return self.transform(self._get(i))
+            if self._transform is None:
+                return self._get(i)
+            else:
+                return self._transform(self._get(i))
 
     def _get(self, i: int) -> dict:
         if i > len(self):
             raise ValueError(f"Index {i} is out of bounds")
-        number = self.image_numbers[i]
+        number = self._image_numbers[i]
         info = None
         if self._needs_info:
-            info_file = self.root / f"{number}.{_Extension.INFO}"
+            info_file = self._root / f"{number}.{_Extension.INFO}"
             with info_file.open("r") as f:
                 info = json.load(f)
 
         ret = {}
-        for modality in self.modalities:
+        for modality in self._modalities:
             ret[modality] = self._open_modality(modality, number, info)
```

Comment:
I would rather store `open` logic within the modality object in case we desire to extend them later

Reply:
I didn't want to expose modality logic to the user, so I decided to go with a single enum.
I'm thinking about classes, it looks more extensible and probably the way to go in the future, however, providing a list of class names looks a bit worse than a list of enums.

Reply:
Probably enum of classes is the way to go =)